### ISSUE TITLE
Error when retrieving credentials from iam-role: Credential refresh failed, response did not contain: access_key, secret_key, token, expiry_time

### ISSUE THREAD

**Issue by kylegalbraith** [boto/botocore#1617]
We are seeing a strange issue relating to `boto3` and `botocore`. The following error is being thrown sporadically when we try to read from S3 or utilize an SQS client.

`Error when retrieving credentials from iam-role: Credential refresh failed, response did not contain: access_key, secret_key, token, expiry_time`

It appears that the credentials are not correctly getting refreshed via the assumed IAM role. This a Python application running inside of a Docker container within EKS. An example piece of code is below.

```python
def fetch_message(s3, bucket, key):
    response = s3.get_object(Bucket=bucket, Key=key)
```

Does anybody have any ideas why this is happening and whether or not this is a known issue with boto?

**Comment by JordonPhillips** [comment]
It looks like you're sourcing credentials from the EC2 Instance Metadata and the request to fetch them failed. By default we don't retry those requests, but you can add retries with `metadata_service_num_attempts` and `metadata_service_timeout` in the config file ([docs](https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#credentials)).

**Comment by kylegalbraith** [comment]
@JordonPhillips Thank you for the response, so would this be a matter of adding those environment variables into the container this is happening in?

**Comment by TattiQ** [comment]
@JordonPhillips  what if  we are sourcing creds from a k8s pod which runs on a k8s worker node (ec2 instance) , so not directly on EC2 instance.  do we set AWS_METADATA_SERVICE_NUM_ATTEMPTS as an env var to a pod? is it still legit then?  thanks!

**Comment by shshe** [comment]
I'm also using kube2iam to have a pod assume an IAM role and seeing this error sporadically. It sometimes happens at the start of the container, but we've also seen it happen after the containers been running for a while. Any suggestions on workarounds? We've set `AWS_METADATA_SERVICE_NUM_ATTEMPTS` but it seems to have no effect.

**Comment by TattiQ** [comment]
@shshe what does botocore debug log level say ? also, do you use celery?

**Comment by shshe** [comment]
Hi @TattiQ , unfortunately, we didn't have DEBUG level turned on. I think our issue lies in kube2iam introducing latency when querying the EC2 metadata URI. Here's the related issue:

https://github.com/jtblin/kube2iam/issues/31

We're currently trying the workaround with setting `AWS_METADATA_SERVICE_NUM_ATTEMPTS` and increasing `AWS_METADATA_SERVICE_TIMEOUT`. So far, the issue hasn't cropped up yet.

**Comment by spinus** [comment]
Hey guys, did you figure out this issue by any chance?

**Comment by swetashre** [comment]
Following up on this issue. Solution provided by @JordonPhillips [here](https://github.com/boto/botocore/issues/1617#issuecomment-446385055) should fix the issue. Is anyone still getting the error ? if yes, please reopen a new issue. I would be happy to help.

**Comment by chrisbrownnwyc** [comment]
We're having this issue specifically with K8S as well. Did setting the Environment variables works?

**Comment by kylegalbraith** [comment]
> We're having this issue specifically with K8S as well. Did setting the Environment variables work?

Yes, setting the environment variables (specifically the retry attempts) seems to have mostly resolved the issue for us. We are also in an EKS K8s environment.

**Comment by dingobar** [comment]
Hi, I was facing this issue running python in a pod in an EKS cluster, and it seems at first glance the retries/timeout solution worked. Did anyone figure out a reason why these requests fail? I've seen pods restart hundreds of times because of this and I'm curious if there is something in the EKS setup that can be used to mitigate.

**Comment by sethatron** [comment]
Bump on this, I am also seeing this issue in kube2iam/EKS

**Comment by alirezaDavid** [comment]
This is also happening for us. (python ---> kube2iam ---> AWS)
We switch to the service account (native EKS way of authenticating) because of this matter.

**Comment by arvindsree** [comment]
Seems to be the error return while handling this issue https://github.com/boto/boto3/issues/1751. The workaround when we hit this issue was to re-attach the instance metadata.
Increasing the retry attempts did not do anything

**Comment by madina-iz** [comment]
I ran into this issue while working in a Jupyter notebook on EC2 instance.  When it first started happening this month, all I had to do is rerun the code and it would work again on the second or third try. However, additional attempts stopped working for me this week. After struggling with a few different options which didn't work in my case, I finally decided to upgrade my Python (from 3.6 to 3.7) and dask (from 2021.11.01 to 2022.2.0), and that fixed the issue for me completely.
